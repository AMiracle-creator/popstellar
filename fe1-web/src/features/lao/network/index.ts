import { ActionType, MessageRegistry, ObjectType } from 'core/network/jsonrpc/messages';

import {
  handleLaoCreateMessage,
  handleLaoStateMessage,
  handleLaoUpdatePropertiesMessage,
} from './LaoHandler';
import { CreateLao, StateLao, UpdateLao } from './messages';

/**
 * Configures the network callbacks in a MessageRegistry.
 *
 * @param registry - The MessageRegistry where we want to add the mappings
 */
export function configureNetwork(registry: MessageRegistry) {
  registry.add(ObjectType.LAO, ActionType.CREATE, handleLaoCreateMessage, CreateLao.fromJson);
  registry.add(ObjectType.LAO, ActionType.STATE, handleLaoStateMessage, StateLao.fromJson);
  registry.add(
    ObjectType.LAO,
    ActionType.UPDATE_PROPERTIES,
    handleLaoUpdatePropertiesMessage,
    UpdateLao.fromJson,
  );
}
