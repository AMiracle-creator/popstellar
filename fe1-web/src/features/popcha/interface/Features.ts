import { LaoParamList } from '../../../core/navigation/typing/LaoParamList';
import { NavigationDrawerScreen } from '../../../core/navigation/typing/Screen';

export namespace PoPchaFeature {
  export interface LaoScreen extends NavigationDrawerScreen {
    id: keyof LaoParamList;
  }
}
