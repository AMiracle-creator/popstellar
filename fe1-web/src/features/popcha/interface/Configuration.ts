import { Hash } from '../../../core/objects';
import FeatureInterface from '../../../core/objects/FeatureInterface';
import { PoPchaFeature } from './Features';

export const POPCHA_FEATURE_IDENTIFIER = 'popcha';

export interface PoPchaConfiguration {
  // TODO: add the configuration interface
  useCurrentLaoId: () => Hash;
}

export type PoPchaReactContext = Pick<PoPchaConfiguration, 'useCurrentLaoId'>;

export interface PoPchaInterface extends FeatureInterface {
  // TODO: add the interface
  laoScreens: PoPchaFeature.LaoScreen[];
  context: PoPchaReactContext;
}
