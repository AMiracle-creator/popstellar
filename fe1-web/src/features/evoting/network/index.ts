import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash } from 'core/objects';
import { dispatch, getStore } from 'core/redux';

import { EvotingConfiguration } from '../interface';
import { Election } from '../objects';
import { addElection, getElectionById, updateElection } from '../reducer';
import {
  handleCastVoteMessage,
  handleElectionEndMessage,
  handleElectionOpenMessage,
  handleElectionResultMessage,
  handleElectionSetupMessage,
} from './ElectionHandler';
import { CastVote, ElectionResult, EndElection, SetupElection } from './messages';
import { OpenElection } from './messages/OpenElection';

/**
 * Configures the network callbacks in a MessageRegistry.
 * @param configuration - An evoting config object
 */
export const configureNetwork = (configuration: EvotingConfiguration) => {
  // getElectionById bound to the global state
  const boundGetElectionById = (electionId: Hash | string) =>
    getElectionById(electionId, getStore().getState());

  const addElectionEvent = (laoId: Hash | string, election: Election) => {
    const electionState = election.toState();

    dispatch(addElection(electionState));
    dispatch(
      configuration.addEvent(laoId, {
        eventType: Election.EVENT_TYPE,
        id: electionState.id,
        start: electionState.start,
        end: electionState.start,
      }),
    );
  };

  const updateElectionEvent = (election: Election) => {
    const electionState = election.toState();

    dispatch(updateElection(electionState));
    dispatch(
      configuration.updateEvent({
        eventType: Election.EVENT_TYPE,
        id: electionState.id,
        start: electionState.start,
        end: electionState.start,
      }),
    );
  };

  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.SETUP,
    handleElectionSetupMessage(addElectionEvent),
    SetupElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.OPEN,
    handleElectionOpenMessage(boundGetElectionById, updateElectionEvent),
    OpenElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.CAST_VOTE,
    handleCastVoteMessage(boundGetElectionById, updateElectionEvent),
    CastVote.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.END,
    handleElectionEndMessage(boundGetElectionById, updateElectionEvent),
    EndElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.RESULT,
    handleElectionResultMessage(boundGetElectionById, updateElectionEvent),
    ElectionResult.fromJson,
  );
};
