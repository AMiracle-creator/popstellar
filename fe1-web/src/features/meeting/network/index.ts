import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash } from 'core/objects';
import { dispatch, getStore } from 'core/redux';

import { MeetingConfiguration } from '../interface';
import { Meeting, MeetingState } from '../objects';
import { addMeeting, getMeetingById, updateMeeting } from '../reducer';
import { handleMeetingCreateMessage, handleMeetingStateMessage } from './MeetingHandler';
import { CreateMeeting, StateMeeting } from './messages';

/**
 * Configures the network callbacks in a MessageRegistry.
 *
 * @param configuration - The configuration object for the meeting feature
 */
export const configureNetwork = (configuration: MeetingConfiguration) => {
  // getMeetingById bound to the global state
  const boundGetMeetingById = (meetingId: Hash | string) =>
    getMeetingById(meetingId, getStore().getState());

  const addMeetingEvent = (laoId: Hash | string, meetingState: MeetingState) => {
    dispatch(configuration.addEvent(laoId, Meeting.EVENT_TYPE, meetingState.id));
    dispatch(addMeeting(meetingState));
  };

  const updateMeetingEvent = (laoId: Hash | string, meetingState: MeetingState) => {
    dispatch(configuration.addEvent(laoId, Meeting.EVENT_TYPE, meetingState.id));
    dispatch(updateMeeting(meetingState));
  };

  configuration.messageRegistry.add(
    ObjectType.MEETING,
    ActionType.CREATE,
    handleMeetingCreateMessage(addMeetingEvent),
    CreateMeeting.fromJson,
  );

  configuration.messageRegistry.add(
    ObjectType.MEETING,
    ActionType.STATE,
    handleMeetingStateMessage(configuration.getLaoById, boundGetMeetingById, updateMeetingEvent),
    StateMeeting.fromJson,
  );
};
