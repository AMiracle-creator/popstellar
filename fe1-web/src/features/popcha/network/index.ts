import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';

import { PopchaConfiguration } from '../interface';
import { PopchaAuthMsg } from './messages/PopchaAuthMsg';

export const configureNetwork = (configuration: PopchaConfiguration) => {
  configuration.messageRegistry.add(
    ObjectType.POPCHA,
    ActionType.AUTH,
    // We cannot subscribe on this channel, so we should not receive any message
    () => {
      throw new Error(`Received unexpected message on popcha auth channel`);
    },
    PopchaAuthMsg.fromJson,
  );
};
