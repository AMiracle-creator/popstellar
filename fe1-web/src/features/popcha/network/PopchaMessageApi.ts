/**
 * Contains all functions to send popcha related messages.
 * */

import { publish } from 'core/network';
import { Base64UrlData, getPopchaAuthenticationChannel, Hash } from 'core/objects';

import { generateToken } from '../../wallet/objects';
import { PopchaAuthMsg } from './messages/PopchaAuthMsg';

/**
 * Sends a message to the server to authenticate that the user belongs to the lao
 * @param client_id id of the client
 * @param nonce nonce
 * @param popcha_address address of the popcha server
 * @param state state
 * @param laoId id of the lao
 * @returns A promise that resolves when the message has been sent
 */
export const sendPopchaAuthRequest = (
  client_id: Hash,
  nonce: string,
  popcha_address: string,
  state: string,
  laoId: Hash,
): Promise<void> => {
  const token = generateToken(laoId, client_id);
  return token.then((t) => {
    const signedToken = t.privateKey.sign(new Base64UrlData(nonce));
    const popchaChannel = getPopchaAuthenticationChannel(laoId);
    const message = new PopchaAuthMsg({
      client_id: client_id.toString(),
      nonce: nonce,
      identifier: t.publicKey.toString(),
      identifier_proof: signedToken.toString(),
      popcha_address: popcha_address /* to add response_mode */,
      state: state,
    });
    return publish(popchaChannel, message);
  });
};
