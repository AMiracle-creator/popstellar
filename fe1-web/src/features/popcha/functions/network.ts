import { Hash, PopToken, ProtocolError } from 'core/objects';

import { sendPopchaAuthRequest } from '../network/PopchaMessageApi';

/**
 * Get the argument from the url
 * @param url the url
 * @param arg the argument
 * @throws ProtocolError if the argument is missing
 */
const getArg = (url: URL, arg: string) => {
  const value = url.searchParams.get(arg);
  if (!value) {
    throw new ProtocolError(`Missing argument ${arg}`);
  }
  return value;
};

/**
 * Verify the scanned info (url)
 * @param data the scanned data
 * @param laoId the lao id
 * @throws ProtocolError if the scanned data is invalid
 */
const verifyScannedInfo = (data: string, laoId: Hash) => {
  let url: URL;
  try {
    url = new URL(data);
  } catch (e) {
    throw new ProtocolError(`Invalid URL: ${e}`);
  }

  const urlArg = url.searchParams;

  const requiredArguments = [
    'client_id',
    'redirect_uri',
    'login_hint',
    'nonce',
    'response_type',
    'scope',
  ];

  // Check if all required arguments are present
  for (const arg of requiredArguments) {
    if (!urlArg.has(arg)) {
      throw new ProtocolError(`Missing argument ${arg}`);
    }
  }

  // Check if the response respects openid standard
  if (getArg(url, 'response_type') !== 'id_token') {
    throw new ProtocolError('Invalid response type');
  }

  if (!(getArg(url, 'scope').includes('openid') && getArg(url, 'scope').includes('profile'))) {
    throw new ProtocolError('Invalid scope');
  }

  const responseMode = url.searchParams.get('response_mode');

  if (responseMode) {
    if (!(responseMode.includes('query') || responseMode.includes('fragment'))) {
      throw new ProtocolError('Invalid response mode');
    }
  }

  if (getArg(url, 'login_hint') !== laoId.toString()) {
    console.info(`Scanned lao id: ${getArg(url, 'login_hint')}, current lao id: ${laoId}`);
    throw new ProtocolError('Invalid lao id');
  }
};

/**
 * Send an auth request to the server
 * @param data the scanned data (url)
 * @param laoId the current lao id
 * @param generateToken function to deterministically generate a long term token
 * @returns A promise that resolves when the message has been sent
 * @throws ProtocolError if the scanned info is invalid
 * @throws NetworkError if message could not be sent correctly
 */
export const sendAuthRequest = async (
  data: string,
  laoId: Hash,
  generateToken: (laoId: Hash, clientId: Hash | undefined) => Promise<PopToken>,
) => {
  verifyScannedInfo(data, laoId);

  const url = new URL(data);
  const urlArg = url.searchParams;

  return sendPopchaAuthRequest(
    urlArg.get('client_id')!,
    urlArg.get('nonce')!,
    url.host,
    urlArg.get('state'),
    urlArg.get('response_mode'),
    laoId,
    generateToken,
  );
};
