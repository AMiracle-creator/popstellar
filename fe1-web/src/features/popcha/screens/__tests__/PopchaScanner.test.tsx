import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';

import { fireScan } from '__mocks__/expo-camera';
import { mockLaoId } from '__tests__/utils';
import FeatureContext from 'core/contexts/FeatureContext';
import STRINGS from 'resources/strings';

import { POPCHA_FEATURE_IDENTIFIER, PopchaReactContext } from '../../interface';
import { sendPopchaAuthRequest } from '../../network/PopchaMessageApi';
import PopchaScanner from '../PopchaScanner';

const contextValue = {
  [POPCHA_FEATURE_IDENTIFIER]: {
    useCurrentLaoId: () => mockLaoId,
  } as PopchaReactContext,
};

const mockToastShow = jest.fn();
const mockToastRet = {
  show: mockToastShow,
};
jest.mock('react-native-toast-notifications', () => ({
  useToast: () => mockToastRet,
}));

jest.mock('../../network/PopchaMessageApi', () => ({
  sendPopchaAuthRequest: jest.fn(() => Promise.resolve()),
}));

const mockClientId = 'mockClientId';
const mockRedirectUri = 'mockRedirectUri';
const mockLoginHint = mockLaoId.toString();
const mockNonce = 'mockNonce';
const mockResponseType = 'id_token';
const mockScope = 'openid profile';

const mockUrl = new URL('https://valid2.server.example:8000');
mockUrl.searchParams.append('client_id', mockClientId);
mockUrl.searchParams.append('redirect_uri', mockRedirectUri);
mockUrl.searchParams.append('login_hint', mockLoginHint);
mockUrl.searchParams.append('nonce', mockNonce);
mockUrl.searchParams.append('response_type', mockResponseType);
mockUrl.searchParams.append('scope', mockScope);

beforeEach(() => {
  jest.clearAllMocks();
});

/**
 * Test that the scanner shows an error message when the url is invalid
 * @param url an invalid url to test
 */
const testInvalidUrl = async (url: string) => {
  const { getByTestId } = render(
    <FeatureContext.Provider value={contextValue}>
      <PopchaScanner />
    </FeatureContext.Provider>,
  );
  const scannerButton = getByTestId('popcha_scanner_button');
  fireEvent.press(scannerButton);
  fireScan(url);
  await waitFor(() => expect(mockToastShow).toHaveBeenCalledTimes(1));
};

describe('Popcha scanner', () => {
  describe('scanner renders correctly', () => {
    it('renders correctly a closed scanner', () => {
      const { toJSON } = render(
        <FeatureContext.Provider value={contextValue}>
          <PopchaScanner />
        </FeatureContext.Provider>,
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('renders correctly a opened scanner', () => {
      const { getByTestId, toJSON } = render(
        <FeatureContext.Provider value={contextValue}>
          <PopchaScanner />
        </FeatureContext.Provider>,
      );
      const button = getByTestId('popcha_scanner_button');
      fireEvent.press(button);
      expect(toJSON()).toMatchSnapshot();
    });
  });
  describe('scanner verifies correctly', () => {
    it('shows error message with invalid url format', async () => {
      await testInvalidUrl('invalid url');
    });

    it('shows error message with url without client_id', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.delete('client_id');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url without redirect_uri', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.delete('redirect_uri');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url without login_hint', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.delete('login_hint');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url without nonce', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.delete('nonce');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url without response_type', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.delete('response_type');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url without scope', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.delete('scope');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url with invalid scope', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.set('scope', 'invalid');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url with invalid response_type', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.set('response_type', 'scope invalid response type');
      await testInvalidUrl(url.toString());
    });

    it('shows error message with url with invalid response_mode', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.set('response_mode', 'invalid response mode');
      await testInvalidUrl(url.toString());
    });

    it('shows error message when login_hint does not match current laoId', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.set('login_hint', 'invalid login hint');
      await testInvalidUrl(url.toString());
    });
  });

  describe('valid url sends correct response', () => {
    it('sends correct response with valid url', async () => {
      const { getByTestId } = render(
        <FeatureContext.Provider value={contextValue}>
          <PopchaScanner />
        </FeatureContext.Provider>,
      );

      const scannerButton = getByTestId('popcha_scanner_button');
      fireEvent.press(scannerButton);
      fireScan(mockUrl.toString());
      await waitFor(() => expect(mockToastShow).toHaveBeenCalledTimes(0));
    });

    it('does not show error message with valid response mode', async () => {
      const url = new URL(mockUrl.toString());
      url.searchParams.set('response_mode', 'query');
      const { getByTestId } = render(
        <FeatureContext.Provider value={contextValue}>
          <PopchaScanner />
        </FeatureContext.Provider>,
      );

      const scannerButton = getByTestId('popcha_scanner_button');
      fireEvent.press(scannerButton);
      fireScan(url.toString());
      await waitFor(() => expect(mockToastShow).toHaveBeenCalledTimes(0));
    });
  });

  describe('correct behavior when sending request', () => {
    it('closes scanner when it is a successful request', async () => {
      const url = new URL(mockUrl.toString());
      (sendPopchaAuthRequest as jest.Mock).mockReturnValue(Promise.resolve());
      const { getByTestId, getByText, toJSON } = render(
        <FeatureContext.Provider value={contextValue}>
          <PopchaScanner />
        </FeatureContext.Provider>,
      );
      const scannerButton = getByTestId('popcha_scanner_button');
      fireEvent.press(scannerButton);
      fireScan(url.toString());
      await waitFor(() => {
        // wait for scanner to close
        expect(getByText(STRINGS.popcha_open_scanner)).toBeTruthy();
        expect(mockToastShow).toHaveBeenCalledTimes(0);
        expect(toJSON()).toMatchSnapshot();
      });
    });

    it('shows error message with failed request', async () => {
      const url = new URL(mockUrl.toString());
      (sendPopchaAuthRequest as jest.Mock).mockReturnValue(Promise.reject(new Error('error')));
      const { getByTestId, toJSON } = render(
        <FeatureContext.Provider value={contextValue}>
          <PopchaScanner />
        </FeatureContext.Provider>,
      );
      const scannerButton = getByTestId('popcha_scanner_button');
      fireEvent.press(scannerButton);
      fireScan(url.toString());
      await waitFor(() => {
        expect(toJSON()).toMatchSnapshot();
        expect(mockToastShow).toHaveBeenCalledTimes(1);
      });
    });
  });
});
