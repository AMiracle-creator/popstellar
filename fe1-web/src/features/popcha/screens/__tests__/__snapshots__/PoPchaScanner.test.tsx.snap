// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoPcha scanner renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    collapsable={false}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      {
        "padding": 16,
      }
    }
    waitFor={
      [
        {
          "current": null,
        },
      ]
    }
  >
    <View>
      <View>
        <Text
          style={
            {
              "color": "#000",
              "fontSize": 20,
              "lineHeight": 26,
              "marginBottom": 16,
              "textAlign": "left",
            }
          }
        >
          Hello, here is your laoID: 
          -
          M
          2
          5
          9
          c
          5
          J
          l
          a
          k
          L
          V
          l
          X
          B
          R
          a
          8
          9
          3
          p
          c
          E
          -
          4
          s
          7
          s
          s
          O
          8
          J
          v
          5
          g
          X
          O
          v
          t
          r
          e
          E
          =
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
